### DISCUSSION TITLE
Proposal: use interface keyword to automatically wrap concrete classes

### DISCUSSION THREAD

**Discussion by mmkal** [dotnet/csharplang#245] [General]
Interfaces are fine when they represent a contract that could be implemented in many ways. But very often, they will be used to wrap a single concrete class, simply because interfaces work better with Dependency Injection and mocking frameworks.

You'll often see code like this:
```C#
// GreetingPrinter.cs
public class GreetingPrinter
{
  public void Greet(string person) => Console.WriteLine("Hello " + person);
}
// IGreetingPrinter.cs
public interface IGreetingPrinter
{
  public void Greet(string person);
}
// MyClass.cs
public class MyClass
{
  private readonly IGreetingPrinter _greetingPrinter;
  public MyClass(IGreetingPrinter greetingPrinter)
  {
    _greetingPrinter = greetingPrinter;
  }
  public void DoStuff()
  {
    Console.WriteLine("doing stuff");
    _greetingPrinter.Greet("Bob");
  }
}
// MyClassTests.cs
public class MyClassTests
{
  private readonly MyClass _myClass = new MyClass(Mock<IGreetingPrinter>());
  [Test]
  public void DoStuff_Greets()
  {
    _myClass.DoStuff();
    Mock<IGreetingPrinter>().Verify(gp => gp.Greet("Bob"));
  }
}
```

This works, but having to create the `IGreetingPrinter` interface is an unnecessary overhead. People rely on their IDE to do it (extract interface), and it's not always simple to change the signature of a public method without tools like Resharper. Sometimes parameter names can get out of sync, or methods are added to the class but not the interface. In cases like this, it would be preferable to bypass the `IGreetingPrinter` definition completely, and use the `interface` keyword to ask the compiler to *generate* us an interface with the concrete class's public properties and methods:

```C#
// GreetingPrinter.cs
public class GreetingPrinter
{
  public void Greet(string person) => Console.WriteLine("Hello " + person);
}
// MyClass.cs
public class MyClass
{
  private readonly interface<GreetingPrinter> _greetingPrinter;
  public MyClass(interface<GreetingPrinter> greetingPrinter)
  {
    _greetingPrinter = greetingPrinter;
  }
  public void DoStuff()
  {
    Console.WriteLine("doing stuff");
    _greetingPrinter.Greet("Bob");
  }
}
// MyClassTests.cs
public class MyClassTests
{
  private readonly MyClass _myClass = new MyClass(Mock<interface<GreetingPrinter>>());
  [Test]
  public void DoStuff_Greets()
  {
    _myClass.DoStuff();
    Mock<interface<GreetingPrinter>>().Verify(gp => gp.Greet("Bob"));
  }
}
```

The compiler does the boring, easily automate-able work for us of creating the interface. It doesn't clutter up our file system or the class definition file. It could even be used to define new classes implementing the same contract:
```C#
public class EnthusiasticGreetingPrinter : interface<GreetingPrinter>
{
  public void Greet(string person) => Console.WriteLine("HELLO " + person.ToUpper());
}
```
 The `interface` keyword is already reserved, so this addition should be backwards compatible.

Thoughts? Could this work?

**Comment by sharwell** [comment]
This proposal makes me uncomfortable, primarily because it appears to primarily assist in writing ~~poor code~~ code that is *difficult to maintain*. I see the following potential scenarios where it could be leveraged:

### The interface is defined so tests can use a mock implementation

This scenario should be avoided. Mock implementations are difficult to write and debug and result in tests that are both fragile and do not reflect the behavior of production code. Use instances of the concrete object instead. If the concrete object is not able to run in a test environment, a new concrete implementation of the interface can be defined in test code to replace the original. Where mocking frameworks result in code that looks like a *use* of the interface, this approach results in code that looks like an *implementation* of the interface, which is much easier to locate and debug. The proposal doesn't appear to address the latter strategy, and could lead to confusion regarding it. In addition, the latter strategy is fairly well supported by the use of virtual methods in the original implementation.

### The interface is defined as part of an API boundary

This scenario (and interface with one implementation) is acceptable, but for the sake of providing a stable API and communicating it is preferable to define the interface separately. One advantage provided by interfaces at API boundaries is external consumers should not need to be exposed to the implementation details in order to understand the interface. The proposal described here necessarily interleaves the two, adding complexity to a design intended to avoid the complexity.

**Comment by HaloFour** [comment]
@sharwell 

Agreed.  Having the interface separate forces thinking about the contract distinctly from the implementation.  The interface is significantly more important than any concrete class that implements it and should be separate.  Trying to modify that contract should require additional effort, not be as simple as just modifying the signature of some method on the class.

**Comment by mmkal** [comment]
Thanks for the response @sharwell 

I agree that this wouldn't be as appropriate for interfaces defined as part of an API boundary - it's clearer if external consumers use `IGreetingPrinter` than `interface<GreetingPrinter>`. But that would be a design decision for the team writing the API. There are myriad ways to make a bad API.

But about your mock implementation point, I'm not sure I understand. I'm not suggesting manually writing mock implementations - isn't it common to unit test using a mocking framework like [Moq](https://github.com/moq/moq4)? Moq, Rhinomoq, and other similar frameworks _can't_ mock concrete classes. So currently we have two choices:

1. don't mock, and make unit tests very difficult to setup in an isolated, safe, and fast way.
2. mock by first writing an interface whose only purpose is to wrap the concrete class you've written.

Are you suggesting that we should be doing the first one?

I also feel that the fact that Resharper/Visual Studio's `right click -> extract interface` tools exist slightly prove that this is something that very often isn't work appropriate for a human, but a compiler.

**Comment by mmkal** [comment]
@HaloFour you are absolutely right, for many interfaces - and those should continue to be written as they are right now. I'm not in any way suggesting that we move away from interfaces in general. They are invaluable for defining an externally-facing contract, when that's what you want to do.

But very often, possibly even in the majority of .NET projects, classes are written for something like an ASP.NET stack, and are not open sourced, put on NuGet or otherwise externally consumed. Many of these classes will never be referenced outside of the solution they live in. Why would you want modifying the contract of those classes to require more effort than is necessary? Any "breaking" changes to the contract would be immediately caught by the compiler/intellisense, and quickly fixed by the developer.

**Comment by sharwell** [comment]
@mmkal About 18 months ago the team I'm on faced a similar challenge to what you describe as the first option. The motivating factor was the use of mocks was causing us major problems:

1. When changing interface semantics slightly (e.g. throw exception instead of returning `null`), it was *extremely* difficult to locate the mocked implementation(s) of that method to ensure they followed the new semantics. This was multiplied over by the use of auto-mocking frameworks in some parts of the tests. Every change had to be implemented at least once in production code, and any number of times in the mocking framework's DSL.
2. We frequently encountered false-positive failures: one or more tests would fail even though production code behaved properly for the same operation.
3. We frequently encountered false-negative failures: one or more tests would pass even though the production code behaved improperly for the same operation.
4. When a test failed, it was difficult to understand whether the failure was a bug in code or a bug in tests. It was also difficult to understand the set of actions a user could take to observe that bug in a production scenario.

We started by challenging the assertion that writing tests without mocks was difficult. Our sprints were already proving that mocks were a major barrier to maintaining code, so we just needed to find an approach that wasn't as hard. The approach we went with involved the following:

1. Write interfaces for modules which communicate with the outside world (UI, file system, web services, external tools, etc.).
2. By default, set up tests using the production implementations of all interfaces. Replace only the minimal subset necessary to account for limitations in the test environment (e.g. a command line tool produced by another team isn't available, so we replace the process runner for this tool with a fake). Make the replacement a new concrete implementation as opposed to a mock so it can be easily located, debugged, and shared between tests that depend on it.
3. Instead of writing tests for individual methods in code, write tests from the perspective of an outside user. The tests declare the behavior of the user/UI, file system, web services, external tools, etc., and then take one or more actions, and finally observe the outcome from the perspective of another external system.
4. When testing subsystems, the "outside world" is redefined as the interface boundary for that subsystems, so the test inputs/outputs are observed at that boundary layer.
5. To ensure we're covering all implementation details we use code coverage analysis of our test runs. These tools do an even better job than method-level unit testing at ensuring edge cases are accounted for in tests.

Our findings were overwhelmingly positive. We not only addressed the maintainability problem we were facing with mocks, but we also overcame other things:

* Tests were at least as fast as they were before. Instantiation of concrete objects is faster than creation of mock implementations.
* Tests were at least as easy to write as they were before. In many cases they were easier - implementations were now readily available and correct, and debugging worked as expected.
* False positives and negatives in testing were reduced.
* Since our assertions were written at module boundaries, test failures were almost always easily described in terms of "If a user does X (and perhaps the system has Y occur), they will observe incorrect behavior R".

Today we've essentially abandoned the use of mocks (while reserving the ability to use them if we ever find a case where they really are required). On the way to this decision, all of our coverage metrics increased, along with our agility and confidence in the test suite to reveal bugs before shipping. In addition our tests run notably faster (more than 20% improvement) and are now deterministic.

**Comment by YaakovDavis** [comment]
I recently performed an analysis on a codebase I'm working in; 70% of the interfaces had only 1 implementation, and ~20% had only 2 implementations.

The redundant interfaces only complicate maintenance, while creating a false illusion of "extensibility". A bad abstraction wrapped in an interface is still a bad abstraction.

**Comment by mmkal** [comment]
@YaakovDavis - interesting, I suspect this is pretty common. Do you have a shareable tool for getting these stats?

@sharwell isn't what you're describing integration testing, not unit testing? If that philosophy works for your team that's great. But unit testing is still a thing.

Plus, with Moq, for example, "finding and updating the mock implementation" would be as simple as finding/changing this line, in the test that is newly failing since the behaviour change:
```C#
Mock<interface<GreetingPrinter>>()
    .Setup(g => g.Greet())
    .Throws<InvalidOperationException>();
```

I do understand your point, but this proposal, to my mind, would deal with a real problem that most .NET projects I've worked on suffer from: codebases bloated with interfaces that provide very little value. 

I completely respect your viewpoint, but my problem is "interfaces are taking up too much of my filesystem, time and mental energy". It'd be a little disappointing if the reason for this feature not going ahead was "you shouldn't be unit testing/using Moq in the first place" - even if that's a valid opinion. It's just unrealistic to expect unit testing to go away any time soon.

**Comment by YaakovDavis** [comment]
@mmkal 
I don't have access to it at this right moment, but I'll try locating it next week.

**Comment by YaakovDavis** [comment]
@mmkal 
> interfaces that provide very little value

I'd add that beyond not *adding* value, they often take away value, by hiding information. 

Often times, the proper functioning of a system relies on a specific, singular, **concrete** implementation. In such cases, there's no point in hiding the implementation behind an interface.

Hiding a singular artifact, is not *abstraction*, but *obfuscation*.

**Comment by sharwell** [comment]
@YaakovDavis That depends heavily on your viewpoint. If you are on the team responsible for the singular concrete implementation, then that's correct. However, if you are on a team that simply consumes the implementation then the implementation details hurt in two ways:

1. They make it more difficult for consumers to get an overview of the intentionally exposed API (noise)
1. They make it too easy for implementations to require consumers know the implementation details in order to use the API. Cross-team APIs should have clearly specified behavior and avoid the requirement to understand implementation details, in order to promote independence and agility without the added risk of breaking downstream consumers.

**Comment by YaakovDavis** [comment]
@sharwell 
API friendliness doesn't depend on the API being composed of interfaces.
The majority of Winforms, WPF, & other well-established .Net libraries is composed of concrete classes that do what you expect them to.

Class/member visibility is a basic facility in enabling this.

Interfaces are helpful when you have multiple implementations, or expect to have such multiplicity in the future for obvious reasons. In many codebases, they're used merely as pointless facades over concrete, singular implementations.

**Comment by sharwell** [comment]
@mmkal I consider our tests (at least the ones I'm talking about here) to be unit tests. It helped to describe desired characteristics of individual tests, such as the following:

* Failure of a well-written test should mean a change occurred for which a user can observe a change in the application behavior under some scenario.
* If a well-written test fails, a developer should be able to describe a user scenario under which a change in behavior can be observed.
* If a developer changes application code such that the application behavior does not change, no test should start to fail unless that test is specifically designated as a performance test.

In order to meet these in our unit tests, we found it particularly valuable to structure the tests to mimic user actions as opposed to being focused on calling specific methods.

**Edit:** If tests are structured in a way that related to behavior of a user or the system¹, it's much easier to communicate the testing strategy to a QA team, and it's much easier to write and verify regression tests that match reports made by users.

¹ File system permissions is one example of system behavior.

> In many codebases, they're used merely as pointless facades over concrete, singular implementations.

I agree. These days we only start with an interface if we're working on an API layer intended to expose stable functionality to unknown code in different components. Inside of the implementation of some functional component, we avoid interfaces. One premise of our design is any component can use the functionality exposed in the designated API of any other component at any point, and the implementation is not allowed to make assumptions about how the API is used aside from the explicitly-documented restrictions for the API methods (e.g. pre- and post-conditions).

**Comment by mmkal** [comment]
@sharwell maybe it would be helpful to separate out the reasons interfaces are currently used:

1. To expose a contract to external consumers, independent of implementation detail.
1. To abstract multiple implementations of a contract.
1. To allow for easy dependency injection and mocking.

For 1 and 2 - the current setup is pretty much fine. 
But I maintain that this feature would significantly increase productivity on projects heavily relying on 3 as a pattern. And there really are a lot of those projects. Is there any chance this feature could be considered for those projects?

Here's one other useful real-world scenario this could be useful:

When consuming an external library, the API is not always designed the way your team would like. As an example which has hit me IRL: the [Microsoft Cognitive Services Emotion API](https://github.com/Microsoft/Cognitive-Emotion-Windows/). For whatever reason, the MS development team decided to make the [`IEmotionServiceClient`](https://github.com/Microsoft/Cognitive-Emotion-Windows/blob/master/ClientLibrary/IEmotionServiceClient.cs) interface `internal`. Only the concrete implementation, [`EmotionServiceClient`](https://github.com/Microsoft/Cognitive-Emotion-Windows/blob/master/ClientLibrary/EmotionServiceClient.cs) was exposed. This meant, in order to isolate it and not actually make HTTP requests when testing, we created an otherwise pointless wrapper class, which implemented an interface with exactly the same contract as Microsoft's internal one. It would have been great if instead we could have just used `interface<EmotionServiceClient>` and had much less code to maintain. How would that have fit into your strategy of using concrete classes whenever possible?

**Comment by HaloFour** [comment]
This could be easily accomplished by a source generator:

```cs
[AutoInterface]
public partial class GreetingPrinter {
    public void Greet(string person) => Console.WriteLine("Hello " + person);
}
```

The generator could then generate the following:

```cs
public interface IGreetingPrinter {
    void Greet(string person);
}

public partial class GreetingPrinter : IGreetingPrinter { }
```

**Comment by mmkal** [comment]
@HaloFour that is true, and something I have looked into - but the same could be said of many language features. I suppose I'll have to settle for that unless others' opinions somehow change.
